jax/_src/random.py

```diff
@@ -1562,32 +1568,52 @@ def categorical(key: ArrayLike,
     shape: Optional, a tuple of nonnegative integers representing the result shape.
       Must be broadcast-compatible with ``np.delete(logits.shape, axis)``.
       The default (None) produces a result shape equal to ``np.delete(logits.shape, axis)``.
+    replace: If True, perform sampling without replacement. Default (False) is to
+      perform sampling with replacement.
 
   Returns:
     A random array with int dtype and shape given by ``shape`` if ``shape``
     is not None, or else ``np.delete(logits.shape, axis)``.
+
+  References:
+    .. [1] Wouter Kool, Herke van Hoof, Max Welling. "Stochastic Beams and Where to Find
+      Them: The Gumbel-Top-k Trick for Sampling Sequences Without Replacement".
+      Proceedings of the 36th International Conference on Machine Learning, PMLR
+      97:3499-3508, 2019. https://proceedings.mlr.press/v97/kool19a.html.
   """
   key, _ = _check_prng_key("categorical", key)
   check_arraylike("categorical", logits)
-  logits_arr = jnp.asarray(logits)
-
-  if axis >= 0:
-    axis -= len(logits_arr.shape)
-
-  batch_shape = tuple(np.delete(logits_arr.shape, axis))
+  logits = jnp.asarray(logits)
+  axis = canonicalize_axis(axis, logits.ndim)
   if shape is None:
-    shape = batch_shape
+    shape = tuple(np.delete(logits.shape, axis))
   else:
     shape = core.canonicalize_shape(shape)
-    _check_shape("categorical", shape, batch_shape)
-
-  shape_prefix = shape[:len(shape)-len(batch_shape)]
-  logits_shape = list(shape[len(shape) - len(batch_shape):])
-  logits_shape.insert(axis % len(logits_arr.shape), logits_arr.shape[axis])
-  return jnp.argmax(
-      gumbel(key, (*shape_prefix, *logits_shape), logits_arr.dtype) +
-      lax.expand_dims(logits_arr, tuple(range(len(shape_prefix)))),
-      axis=axis)
+    _check_shape("categorical", shape, np.delete(logits.shape, axis))
+
+  batch_ndim = len(shape) - (logits.ndim - 1)
+  batch_shape = shape[:batch_ndim]
+
+  if replace:
+    noise = gumbel(key, batch_shape + logits.shape, logits.dtype)
+    logits = lax.broadcast_to_rank(logits, noise.ndim)
+    logits += noise
+    return logits.argmax(batch_ndim + axis)
+
+  else:
+    logits += gumbel(key, logits.shape, logits.dtype)
+    k = math.prod(batch_shape)
+    logits = jnp.moveaxis(logits, axis, -1)
+    try:
+      _, indices = lax.top_k(logits, k)
+    except Exception as e:
```

Comment:
We should not catch a base exception here: if `top_k` raises an unanticipated exception, this approach would obscure the true error.

I'd suggest that before calling `top_k`, we explicitly check the condition `k <= logits.shape[-1]`.